### PR TITLE
Allow automatically re-writing queries 

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
This change is explained in the comments, but it's a feature to keep from needing to manually check every query in auth (or any other plugin). More discussion of this is here: https://github.com/Expensify/Server-Expensify/pull/2646

Essentially, this uses sqlite's `set_authorizer` feature to allow a plugin to register a callback that can say "Oh, you want to do an INSERT into the XYZ table, but we're handling a mocked request? In that case, just run `INSERT INTO XYZ VALUES(-1,2,3);DELETE FROM XYZ WHERE id = -1;` instead".

This prevents having to make a bunch of similar modifications all over the place to do this for every insert or delete.

### CODE REVIEW

**Comment by quinthar** [comment]
Wow clever!

File: sqlitecluster/SQLite.h [hunk 1]
```diff
@@ -85,6 +85,32 @@ class SQLite {
...
+    // followed by a DELETE. When enabled, this is implemented as follows:
+    //
+    // Plugin calls `write`.
+    // 1. If enableRewrite is ON, the rewriteHandler is run as part of the authorizer.
+    // 2. If the rewriteHandler wants to re-write a query, it should return `true` and update the string reference it
+    //    was passed (see setRewriteHandler() below).
+    // 3. If the rewriteHandler returns true, the initial query will fail with SQLITE_AUTH (warnings for this failure
+    //    are suppressed) and the new replacement query will be run in it's place.
+    //
+    // Note: your plugin should only enable rewrites during it's own scope (i.e., it should disable them at the end of
```
Comment (coleaeason) [inline sqlitecluster/SQLite.h]: This seems dangerous, is there a way to force this to be reset at the end of peek/process? Or to attach enableRewrite to a specific command/query and not leave it enabled once that command/query is done?

Comment (tylerkaraszewski) [reply sqlitecluster/SQLite.h]: We could let a particular plugin set it and then enable/disable it for that plugin, that would be a pretty easy change.

Comment (coleaeason) [reply sqlitecluster/SQLite.h]: Lets do that

**Comment by tylerkaraszewski** [comment]
@coleaeason - updated

File: plugins/DB.cpp [hunk 2]
```diff
@@ -3,6 +3,38 @@
...
 #define SLOGPREFIX "{" << _args["-nodeName"] << ":" << getName() << "} "
 
+bool BedrockPlugin_DB::rewriteHandler(int actionCode, const char* table, string& newQuery) {
+// Hack to make logging work in static function
+#undef SLOGPREFIX
+#define SLOGPREFIX "{" << "unknown" << ":" << "DB" << "} "
+    SINFO("bedrock Running DB rewrite handler.");
+    switch (actionCode) {
+        case SQLITE_INSERT:
+            if (table == "chats"s) {
```
Comment (coleaeason) [inline plugins/DB.cpp]: this an expensify specific table, it should go in the expensify plugin that uses it.

File: plugins/DB.cpp [hunk 3]
```diff
@@ -3,6 +3,38 @@
...
+                SINFO("bedrock DB rewrite handler replacing chat insert.");
+                newQuery = "INSERT INTO chats VALUES (-1000, 1, 1, 'fake', 'fake', 'fake'); DELETE FROM chats WHERE chatID = -1000;";
+            } else if (table == "jobs"s) {
```
Comment (coleaeason) [inline plugins/DB.cpp]: this should go in jobs.cpp, no?

File: plugins/DB.cpp [hunk 4]
```diff
@@ -109,10 +141,10 @@ bool BedrockPlugin_DB::processCommand(SQLite& db, BedrockCommand& command) {
 
     // ----------------------------------------------------------------------
     if (SIEquals(request.getVerb(), "Query")) {
-        if (db.getUpdateNoopMode()) {
-            SINFO("Query run in mocked request, just ignoring.");
-            return true;
-        }
+        //if (db.getUpdateNoopMode()) {
```
Comment (coleaeason) [inline plugins/DB.cpp]: why is this commented out instead of deleted?

File: sqlitecluster/SQLite.cpp [hunk 5]
```diff
@@ -444,6 +446,7 @@ string SQLite::read(const string& query) {
 
 bool SQLite::read(const string& query, SQResult& result) {
     uint64_t before = STimeNow();
+    // TODO: Support query re-writing.
```
Comment (coleaeason) [inline sqlitecluster/SQLite.cpp]: Hmmm, why would we need to support rewriting of reads? The caller should be able to re do the read several times and discard the results it doesn't want. I don't think we need to support re-writing these.

**Comment by tylerkaraszewski** [comment]
@coleaeason Updated again.

**Pull request merged**